lib/assert.js

```diff
@@ -931,6 +933,75 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+const mustCallChecks = [];
+
+function runCallChecks(exitCode) {
+  if (exitCode !== 0) return;
+
+  const failed = mustCallChecks.filter((context) => {
+    context.messageSegment = `exactly ${context.exact}`;
+    return context.actual !== context.exact;
+  });
+
+  failed.forEach(function(context) {
+    console.error('Mismatched %s function calls. Expected %s, actual %d.',
+                  context.name,
+                  context.messageSegment,
+                  context.actual);
+    console.error(context.stack.split('\n').slice(2).join('\n'));
+  });
+
+  if (failed.length) process.exit(1);
+}
+
+assert.mustCall = function mustCall(fn, exact = 1) {
+  if (process._exiting)
+    // Tthrow new Error('Cannot use common.mustCall() in process exit handler');
+    throw new ERR_MUSTCALL_UNAVAILABLE_DURING_EXIT();
+  if (typeof fn === 'number') {
+    exact = fn;
+    fn = noop;
+  } else if (fn === undefined) {
+    fn = noop;
+  }
+
+  if (typeof exact !== 'number')
+    throw new ERR_INVALID_ARG_TYPE(`Invalid exact value: ${exact}`);
+
+  // eslint-disable-next-line no-restricted-syntax
+  const stackTrace = util.inspect(new Error());
+
+  const context = {
+    exact: exact,
+    actual: 0,
+    stack: stackTrace,
+    name: fn.name || '<anonymous>'
+  };
+
+  // Add the exit listener only once to avoid listener leak warnings
+  if (mustCallChecks.length === 0) process.on('exit', runCallChecks);
```

Comment:
Currently, `mustCallChecks` can also be empty because of the `.splice()` call below, not just because this is the first call to `assert.mustCall()`.

There’s probably some clever way around this, but I’d probably suggest to keep track of whether the `'exit'` listener is attached through a separate variable.